### PR TITLE
Disable mouse popup for empty space in map selection window

### PR DESCRIPTION
**Pull request by orzm**
Resolves issue https://github.com/ihhub/fheroes2/issues/8738. This has been done in the similar way as for remaining elements of that window: by checking if position of right click area was inside the map list.

Since all of these checks were almost identical, a small refactor has been made by introducing a common function: `ShowIfFound`. It accepts three parameters: reference to ScenarioListBox object containing loaded maps, const reference to fheroes2::Point of mouse right click location, and a reference to function, which should be called if mouse location is in desired area.

https://github.com/user-attachments/assets/a849dec9-9c20-4619-8c2d-d51b321076dd

### CODE REVIEW

File: src/fheroes2/dialog/dialog_selectscenario.cpp [hunk 1]
```diff
@@ -91,14 +91,31 @@ namespace
         fheroes2::showStandardTextMessage( header, body, Dialog::ZERO );
     }
 
-    void mapInfo( const Maps::FileInfo & info )
+    void PlayersToolTip( [[maybe_unused]] const Maps::FileInfo * info = nullptr )
```
Comment (oleg-derevenetz) [inline src/fheroes2/dialog/dialog_selectscenario.cpp:94]: The `[[maybe_unused]]` attribute as useful when variable or parameter may be used _sometimes_ - e.g. in the debug builds only. For cases where the parameter is _never_ used, it is possible to use a more explicit indication of this - just do not name it:

```suggestion
    void PlayersToolTip( const Maps::FileInfo * /* info */ = nullptr )
```

The same applies to the rest of the cases below.

File: src/fheroes2/dialog/dialog_selectscenario.cpp [hunk 2]
```diff
@@ -91,14 +91,31 @@ namespace
...
     {
-        fheroes2::Text header( info.name, fheroes2::FontType::normalYellow() );
+        ShowToolTip( _( "Players Icon" ),
+                     _( "Indicates how many players total are in the scenario. Any positions not occupied by human players will be occupied by computer players." ) );
+    }
+
+    void SizeToolTip( [[maybe_unused]] const Maps::FileInfo * info = nullptr )
```
Comment (oleg-derevenetz) [inline src/fheroes2/dialog/dialog_selectscenario.cpp:100]:
```suggestion
    void SizeToolTip( const Maps::FileInfo * /* info */ = nullptr )
```

File: src/fheroes2/dialog/dialog_selectscenario.cpp [hunk 3]
```diff
@@ -91,14 +91,31 @@ namespace
...
+    {
+        ShowToolTip( _( "Size Icon" ), _( "Indicates whether the map\nis small (36 x 36), medium\n(72 x 72), large (108 x 108),\nor extra large (144 x 144)." ) );
+    }
+
+    void MapTypeToolTip( [[maybe_unused]] const Maps::FileInfo * info = nullptr )
```
Comment (oleg-derevenetz) [inline src/fheroes2/dialog/dialog_selectscenario.cpp:105]:
```suggestion
    void MapTypeToolTip( const Maps::FileInfo * /* info */ = nullptr )
```

**Review by oleg-derevenetz — COMMENTED** [review]
Hi @orzm I just left a few (In fact, just one) suggestion, could you please take a look when you have time?

File: src/fheroes2/dialog/dialog_selectscenario.cpp (hunk 1 above)
Comment (orzm) [reply src/fheroes2/dialog/dialog_selectscenario.cpp]: Applied suggested changes here and for remaining cases

File: src/fheroes2/dialog/dialog_selectscenario.cpp [hunk 4]
```diff
@@ -208,6 +226,15 @@ namespace
...
     }
+
+    template <class ShowFunction>
+    void ShowIfFound( ScenarioListBox & listbox, const fheroes2::Point & mouseLocation, ShowFunction && function )
+    {
+        const Maps::FileInfo * item = listbox.GetFromPosition( mouseLocation );
+        if ( item ) {
+            std::forward<ShowFunction>( function )( item );
+        }
+    }
```
Comment (oleg-derevenetz) [inline src/fheroes2/dialog/dialog_selectscenario.cpp:237]: In this case, there is no need to use a [universal reference](https://isocpp.org/blog/2012/11/universal-references-in-c11-scott-meyers), using it here does not give anything (that's why all these complaints about missing `std::forward`).

```suggestion
    template <class ShowFunction>
    void ShowIfFound( ScenarioListBox & listbox, const fheroes2::Point & mouseLocation, const ShowFunction & function )
    {
        const Maps::FileInfo * item = listbox.GetFromPosition( mouseLocation );
        if ( item ) {
            function( item );
        }
    }
```

File: src/fheroes2/dialog/dialog_selectscenario.cpp [hunk 5]
```diff
@@ -631,43 +658,41 @@ const Maps::FileInfo * Dialog::SelectScenario( const MapsFileInfoList & allMaps,
         else if ( le.isMouseRightButtonPressedInArea( buttonSelectAll.area() ) ) {
             ShowToolTip( _( "All Maps" ), _( "View all maps, regardless of size." ) );
         }
-        else if ( le.isMouseRightButtonPressedInArea( countPlayers ) || le.isMouseRightButtonPressedInArea( curCountPlayer ) ) {
-            ShowToolTip( _( "Players Icon" ),
-                         _( "Indicates how many players total are in the scenario. Any positions not occupied by human players will be occupied by computer players." ) );
+        else if ( le.isMouseRightButtonPressedInArea( countPlayers ) ) {
+            ShowIfFound( listbox, le.getMouseCursorPos(), []( const Maps::FileInfo * info ) { PlayersToolTip( info ); } );
```
Comment (oleg-derevenetz) [inline src/fheroes2/dialog/dialog_selectscenario.cpp:662]: There (and in the rest of the similar cases below) is no need in lambda. Just pass the (pointer to) function directly:

```suggestion
            ShowIfFound( listbox, le.getMouseCursorPos(), PlayersToolTip );
```

File: src/fheroes2/dialog/dialog_selectscenario.cpp [hunk 6]
```diff
@@ -631,43 +658,41 @@ const Maps::FileInfo * Dialog::SelectScenario( const MapsFileInfoList & allMaps,
...
+        }
+        else if ( le.isMouseRightButtonPressedInArea( curCountPlayer ) ) {
+            PlayersToolTip();
+        }
+        else if ( le.isMouseRightButtonPressedInArea( sizeMaps ) ) {
+            ShowIfFound( listbox, le.getMouseCursorPos(), []( const Maps::FileInfo * info ) { SizeToolTip( info ); } );
```
Comment (oleg-derevenetz) [inline src/fheroes2/dialog/dialog_selectscenario.cpp:668]:
```suggestion
            ShowIfFound( listbox, le.getMouseCursorPos(), SizeToolTip );
```

File: src/fheroes2/dialog/dialog_selectscenario.cpp [hunk 7]
```diff
@@ -631,43 +658,41 @@ const Maps::FileInfo * Dialog::SelectScenario( const MapsFileInfoList & allMaps,
...
-        else if ( le.isMouseRightButtonPressedInArea( sizeMaps ) || le.isMouseRightButtonPressedInArea( curMapSize ) ) {
-            ShowToolTip( _( "Size Icon" ), _( "Indicates whether the map\nis small (36 x 36), medium\n(72 x 72), large (108 x 108),\nor extra large (144 x 144)." ) );
+        else if ( le.isMouseRightButtonPressedInArea( curMapSize ) ) {
+            SizeToolTip();
         }
-        else if ( le.isMouseRightButtonPressedInArea( mapTypes ) || le.isMouseRightButtonPressedInArea( curMapType ) ) {
-            ShowToolTip( _( "Map Type" ),
-                         _( "Indicates whether the map is made for \"The Succession Wars\", \"The Price of Loyalty\" or \"Resurrection\" version of the game." ) );
+        else if ( le.isMouseRightButtonPressedInArea( mapTypes ) ) {
+            ShowIfFound( listbox, le.getMouseCursorPos(), []( const Maps::FileInfo * info ) { MapTypeToolTip( info ); } );
```
Comment (oleg-derevenetz) [inline src/fheroes2/dialog/dialog_selectscenario.cpp:674]:
```suggestion
            ShowIfFound( listbox, le.getMouseCursorPos(), MapTypeToolTip );
```

File: src/fheroes2/dialog/dialog_selectscenario.cpp [hunk 8]
```diff
@@ -631,43 +658,41 @@ const Maps::FileInfo * Dialog::SelectScenario( const MapsFileInfoList & allMaps,
...
+        }
+        else if ( le.isMouseRightButtonPressedInArea( curMapType ) ) {
+            MapTypeToolTip();
         }
         else if ( le.isMouseRightButtonPressedInArea( mapNames ) ) {
-            const Maps::FileInfo * item = listbox.GetFromPosition( le.getMouseCursorPos() );
-            if ( item ) {
-                mapInfo( *item );
-            }
+            ShowIfFound( listbox, le.getMouseCursorPos(), []( const Maps::FileInfo * info ) { mapInfo( info ); } );
```
Comment (oleg-derevenetz) [inline src/fheroes2/dialog/dialog_selectscenario.cpp:680]:
```suggestion
            ShowIfFound( listbox, le.getMouseCursorPos(), mapInfo );
```

File: src/fheroes2/dialog/dialog_selectscenario.cpp [hunk 9]
```diff
@@ -631,43 +658,41 @@ const Maps::FileInfo * Dialog::SelectScenario( const MapsFileInfoList & allMaps,
...
         }
         else if ( le.isMouseRightButtonPressedInArea( curMapName ) ) {
             ShowToolTip( _( "Selected Name" ), _( "The name of the currently selected map." ) );
         }
         else if ( le.isMouseRightButtonPressedInArea( victoryConds ) ) {
-            const Maps::FileInfo * item = listbox.GetFromPosition( le.getMouseCursorPos() );
-            if ( item ) {
-                VictoryConditionInfo( *item );
-            }
+            ShowIfFound( listbox, le.getMouseCursorPos(), []( const Maps::FileInfo * info ) { VictoryConditionInfo( info ); } );
```
Comment (oleg-derevenetz) [inline src/fheroes2/dialog/dialog_selectscenario.cpp:686]:
```suggestion
            ShowIfFound( listbox, le.getMouseCursorPos(), VictoryConditionInfo );
```

File: src/fheroes2/dialog/dialog_selectscenario.cpp [hunk 10]
```diff
@@ -631,43 +658,41 @@ const Maps::FileInfo * Dialog::SelectScenario( const MapsFileInfoList & allMaps,
...
         }
         else if ( le.isMouseRightButtonPressedInArea( lossConds ) ) {
-            const Maps::FileInfo * item = listbox.GetFromPosition( le.getMouseCursorPos() );
-            if ( item ) {
-                LossConditionInfo( *item );
-            }
+            ShowIfFound( listbox, le.getMouseCursorPos(), []( const Maps::FileInfo * info ) { LossConditionInfo( info ); } );
```
Comment (oleg-derevenetz) [inline src/fheroes2/dialog/dialog_selectscenario.cpp:689]:
```suggestion
            ShowIfFound( listbox, le.getMouseCursorPos(), LossConditionInfo );
```

File: src/fheroes2/dialog/dialog_selectscenario.cpp [hunk 11]
```diff
@@ -27,6 +27,7 @@
 #include <cassert>
 #include <cstddef>
 #include <string>
+#include <utility>
```
Comment (oleg-derevenetz) [inline src/fheroes2/dialog/dialog_selectscenario.cpp:30]: No need for `std::forward` - no need for this header:

```suggestion
```

**Review by oleg-derevenetz — COMMENTED** [review]
Hi @orzm I just left a little more suggestions to simplify the code. Could you take a look at them when you find the time?

File: src/fheroes2/dialog/dialog_selectscenario.cpp (hunk 4 above)
Comment (orzm) [reply src/fheroes2/dialog/dialog_selectscenario.cpp]: Done. Thanks for linked article, it really helps to know the difference between rvalue and universal references.

File: src/fheroes2/dialog/dialog_selectscenario.cpp (hunk 5 above)
Comment (orzm) [reply src/fheroes2/dialog/dialog_selectscenario.cpp]: Done

File: src/fheroes2/dialog/dialog_selectscenario.cpp (hunk 6 above)
Comment (orzm) [reply src/fheroes2/dialog/dialog_selectscenario.cpp]: Done

File: src/fheroes2/dialog/dialog_selectscenario.cpp (hunk 7 above)
Comment (orzm) [reply src/fheroes2/dialog/dialog_selectscenario.cpp]: Done

File: src/fheroes2/dialog/dialog_selectscenario.cpp (hunk 8 above)
Comment (orzm) [reply src/fheroes2/dialog/dialog_selectscenario.cpp]: Done

File: src/fheroes2/dialog/dialog_selectscenario.cpp (hunk 9 above)
Comment (orzm) [reply src/fheroes2/dialog/dialog_selectscenario.cpp]: Done

File: src/fheroes2/dialog/dialog_selectscenario.cpp (hunk 10 above)
Comment (orzm) [reply src/fheroes2/dialog/dialog_selectscenario.cpp]: Done

File: src/fheroes2/dialog/dialog_selectscenario.cpp (hunk 11 above)
Comment (orzm) [reply src/fheroes2/dialog/dialog_selectscenario.cpp]: Removed

**Review by oleg-derevenetz — APPROVED** [review]

**Comment by oleg-derevenetz** [comment]
@orzm Thank you very much!

**Pull request merged**